### ISSUE TITLE
Support for hardcoded kernel gradients

### ISSUE THREAD

**Issue by henrymoss** [GPflow/GPflow#1539]
Hi all

I have been using GPflow with a custom kernel for string inputs and have an interesting (although quite niche) dilemma.

The calculation of my kernel  require the solving of dynamic programs. Unfortunately, this means that Tensorflow's autodiff cannot efficiently calculate the kernel's gradients w.r.t its hyper-parameters (without exponentially large memory use). In my GPy implementation, I efficiently calculate kernel gradients using additional dynamic programs. 

To allow the use of hardcoded gradients when fitting my GP, I have had to write a custom training wrapper that sidesteps auto-diff (similar to the way that kernel parameters are optimised in GPy). 

I wonder if there is a more native way to allow gpflow to use hard-coded gradients? Down the line, I plan on having kernels with some kernel parameters than can be auto-differentiated and some that cannot, and so a more robust solution would be needed.

Thanks

Henry

**Comment by st--** [comment]
Hi @henrymoss, a native way for custom gradients is in essence a TensorFlow question, not a gpflow one. You can use tf.custom_gradient in conjunction with tf.py_function to write an operation that has a gradient computed any way you like. Hope that helps. If you'd like a longer answer, this is more of a stackoverflow question so in that case please post there (#gpflow/#tensorflow tags)!

**Comment by henrymoss** [comment]
@st--  Thanks for the quick response! 

I will take a look at the custom_gradients.

Henry
